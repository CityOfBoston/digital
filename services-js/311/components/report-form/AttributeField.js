// @flow

import React from 'react';
import { css } from 'glamor';
import type { ServiceMetadataAttribute } from '../../data/types';

export type Props = {
  attribute: ServiceMetadataAttribute,
};

const TEXT_TEXTAREA_STYLE = css({
  width: '100%',
  height: 180,
});

function renderInformationalAttribute(attribute) {
  return (
    <p key={attribute.code}>{attribute.description}</p>
  );
}

function renderTextAttribute(attribute) {
  return (
    <label key={attribute.code}>
      <p>{attribute.description} {attribute.required ? '(required)' : null}</p>
      <textarea className={TEXT_TEXTAREA_STYLE} />
    </label>
  );
}

function renderPicklistAttribute(attribute) {
  return (
    <label key={attribute.code}>
      <p>{attribute.description}</p>
      <select name={attribute.code}>
        {(attribute.values || []).map(({ key, name }) => <option value={key} key={key}>{name}</option>)}
      </select>
    </label>
  );
}

export default function AttributeField({ attribute }: Props) {
  switch (attribute.type) {
    case 'INFORMATIONAL':
      return renderInformationalAttribute(attribute);
    case 'TEXT':
      return renderTextAttribute(attribute);
    case 'PICKLIST':
      return renderPicklistAttribute(attribute);
    default:
      return null;
  }
}
