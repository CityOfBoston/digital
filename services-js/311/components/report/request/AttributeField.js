// @flow

import React from 'react';
import { action } from 'mobx';
import { observer } from 'mobx-react';
import type Question from '../../../data/store/Question';

export type Props = {
  question: Question;
};

function currentValueAsArray(currentValue): string[] {
  if (!currentValue) {
    return [];
  }

  currentValue = currentValue.slice();

  if (Array.isArray(currentValue)) {
    return currentValue;
  } else {
    return [currentValue];
  }
}

function renderCheckbox(question, onChange) {
  return (
    <label className="cb">
      <input name={question.code} type="checkbox" value="true" className="cb-f" checked={question.value === 'true'} onChange={onChange} />
      <span className="cb-l">{question.description}</span>
    </label>
  );
}

function renderMultiValueListAttribute(question, onChange) {
  const values = currentValueAsArray(question.value);

  return (
    <div>
      <div className="m-v200"><span className="txt-l">{ question.description } {question.required ? '(required)' : null}</span></div>
      { (question.valueOptions || []).map(({ key, name }) => (
        <label className="cb" key={key}>
          <input name={question.code} type="checkbox" value={key} className="cb-f" checked={values.indexOf(key) !== -1} onChange={onChange} />
          <span className="cb-l">{name}</span>
        </label>
      ))
    }
    </div>
  );
}

function renderDatetimeAttribute(question, onChange) {
  return (
    <label className="txt">
      <span className="txt-l">{question.description} {question.required ? '(required)' : null}</span>
      <input type="datetime-local" name={question.code} className="txt-f" value={question.value} onChange={onChange} />
    </label>
  );
}

function renderDateAttribute(question, onChange) {
  return (
    <label className="txt">
      <span className="txt-l">{question.description} {question.required ? '(required)' : null}</span>
      <input type="date" name={question.code} className="txt-f" value={question.value} onChange={onChange} />
    </label>
  );
}

function renderInformationalAttribute(question) {
  return (
    <p className="t--info">{question.description}</p>
  );
}

function renderTextAttribute(question, onChange) {
  return (
    <label className="txt">
      <span className="txt-l">{question.description} {question.required ? '(required)' : null}</span>
      <textarea name={question.code} className="txt-f" value={question.value} onChange={onChange} rows="5" />
    </label>
  );
}

function renderStringAttribute(question, onChange) {
  return (
    <label className="txt">
      <span className="txt-l">{question.description} {question.required ? '(required)' : null}</span>
      <input type="text" name={question.code} className="txt-f" value={question.value} onChange={onChange} />
    </label>
  );
}

function renderNumberAttribute(question, onChange) {
  return (
    <label className="txt">
      <span className="txt-l">{question.description} {question.required ? '(required)' : null}</span>
      <input type="number" name={question.code} className="txt-f" value={question.value} onChange={onChange} />
    </label>
  );
}

function renderSingleValueListAttributeAsRadios(question, onChange) {
  return (
    <div>
      <div className="m-v200"><span className="txt-l">{ question.description } {question.required ? '(required)' : null}</span></div>
      { (question.valueOptions || []).map(({ key, name }) => (
        <label className="ra" key={key}>
          <input name={question.code} type="radio" value={key} className="ra-f" checked={question.value === key} onChange={onChange} />
          <span className="ra-l">{name}</span>
        </label>
        ))
      }
      { !question.required && <label className="ra">
        <input name={question.code} type="radio" value="" className="ra-f" checked={question.value === '' || question.value === null} onChange={onChange} />
        <span className="ra-l">No Answer</span>
        </label>
      }
    </div>
  );
}

export default observer(function AttributeField({ question }: Props) {
  if (!question.visible) {
    return null;
  }

  const onChange = action('onChange', (ev: SyntheticInputEvent) => {
    question.value = ev.target.value;
  });

  const onCheckbox = action('onCheckbox', (ev: SyntheticInputEvent) => {
    question.value = (ev.target.checked || false).toString();
  });

  const onMultivalueList = action('onMultivalueList', (ev: SyntheticInputEvent) => {
    const values = currentValueAsArray(question.value);

    if (ev.target.checked) {
      question.value = [...values, ev.target.value];
    } else {
      question.value = values.filter((v) => v !== ev.target.value);
    }
  });

  switch (question.type) {
    case 'BOOLEAN_CHECKBOX':
      return renderCheckbox(question, onCheckbox);
    case 'INFORMATIONAL':
      return renderInformationalAttribute(question);
    case 'DATETIME':
      return renderDatetimeAttribute(question, onChange);
    case 'DATE':
      return renderDateAttribute(question, onChange);
    case 'STRING':
      return renderStringAttribute(question, onChange);
    case 'NUMBER':
      return renderNumberAttribute(question, onChange);
    case 'TEXT':
      return renderTextAttribute(question, onChange);
    case 'SINGLEVALUELIST':
      return renderSingleValueListAttributeAsRadios(question, onChange);
    case 'MULTIVALUELIST':
      return renderMultiValueListAttribute(question, onMultivalueList);
    default:
      return null;
  }
});
