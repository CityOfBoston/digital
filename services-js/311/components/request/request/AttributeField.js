// @flow

import React from 'react';
import { action } from 'mobx';
import { observer } from 'mobx-react';
import { css } from 'glamor';

import type Question from '../../../data/store/Question';

import AttributeDateField from './AttributeDateField';

export type Props = {|
  question: Question,
|};

const NUMERIC_FIELD_STYLE = css({
  width: '6em',
  textAlign: 'right',
  MozAppearance: 'textfield',
  '::-webkit-outer-spin-button': {
    WebkitAppearance: 'none',
    margin: 0,
  },
  '::-webkit-inner-spin-button': {
    WebkitAppearance: 'none',
    margin: 0,
  },
});

// Returns either one or two lists. If list’s length is less than splitLength,
// returns the entire list as the first element of an array. If list’s length
// is longer, splits in in half and returns each half.
function maybeSplitList(list, splitLength) {
  list = list || [];

  if (list.length < splitLength) {
    return [list];
  } else {
    const firstLength = Math.ceil(list.length / 2);
    return [list.slice(0, firstLength), list.slice(firstLength)];
  }
}

function currentValueAsArray(currentValue): string[] {
  if (!currentValue) {
    return [];
  }

  currentValue = currentValue.slice();

  if (Array.isArray(currentValue)) {
    return currentValue;
  } else {
    return [currentValue];
  }
}

function maybeRenderRequired(required: boolean) {
  if (required) {
    return <span className="t--req">Required</span>;
  } else {
    return null;
  }
}

function renderValidationMessages({
  validationErrorMessages,
  validationInfoMessages,
}: Question) {
  return [
    validationErrorMessages.map((m, i) =>
      <div className="t--subinfo t--err m-t100" key={`err-${i}`}>
        {m}
      </div>
    ),

    validationInfoMessages.map((m, i) =>
      <div className="t--subinfo m-t100" key={`err-${i}`}>
        {m}
      </div>
    ),
  ];
}

function renderCheckbox(question, onChange) {
  return (
    <div className="cb">
      <input
        name={question.code}
        id={question.code}
        type="checkbox"
        value="true"
        className="cb-f"
        checked={question.value === 'true'}
        onChange={onChange}
      />
      <label className="cb-l" htmlFor={question.code}>
        {question.description}
      </label>

      {renderValidationMessages(question)}
    </div>
  );
}

function renderInformationalAttribute(question) {
  return (
    <p className="t--info">
      {question.description}
    </p>
  );
}

function renderTextAttribute(question, onChange) {
  return (
    <div className="txt">
      <label className="txt-l" htmlFor={question.code}>
        {question.description} {maybeRenderRequired(question.required)}
      </label>
      <textarea
        name={question.code}
        id={question.code}
        className="txt-f"
        value={question.value}
        onChange={onChange}
        rows="5"
        aria-required={question.required}
      />

      {renderValidationMessages(question)}
    </div>
  );
}

function renderStringAttribute(question, onChange) {
  const {
    description,
    required,
    code,
    hasSafeValue,
    requirementsMet,
    value,
  } = question;

  const errHighlight = hasSafeValue && !requirementsMet;

  return (
    <div className="txt">
      <label className="txt-l" htmlFor={code}>
        {description} {maybeRenderRequired(required)}
      </label>

      <input
        type="text"
        name={code}
        id={code}
        className={`txt-f ${errHighlight ? 'txt-f--err' : ''}`}
        value={value}
        onChange={onChange}
        aria-required={required}
      />

      {renderValidationMessages(question)}
    </div>
  );
}

function renderNumberAttribute(question, onChange) {
  const {
    description,
    required,
    code,
    hasSafeValue,
    requirementsMet,
    value,
  } = question;

  const errHighlight = hasSafeValue && !requirementsMet;

  return (
    <div className="txt">
      <label className="txt-l">
        {description} {maybeRenderRequired(required)}
      </label>

      <input
        name={code}
        id={code}
        className={`txt-f
          ${NUMERIC_FIELD_STYLE.toString()}
          ${errHighlight ? 'txt-f--err' : ''}`}
        value={value}
        onChange={onChange}
        pattern="[0-9]*"
        aria-required={required}
      />

      {renderValidationMessages(question)}
    </div>
  );
}

function renderMultiValueListAttribute(question, onChange) {
  const { description, required, code, value, valueOptions } = question;

  const values = currentValueAsArray(value);
  const lists = maybeSplitList(valueOptions, 5);

  const labelId = `${code}-label`;

  return (
    <div role="group" aria-labelledby={labelId}>
      <div className="m-v300">
        <label className="txt-l" id={labelId}>
          {description} {maybeRenderRequired(required)}
        </label>
      </div>

      <div className="g">
        {lists.map((list, i) =>
          <div
            className={lists.length === 1 ? 'g--12' : 'g--6'}
            key={`list-${i}`}
            style={{ marginTop: '-.5rem' }}
          >
            {list.map(({ key, name }) =>
              <div className="cb" key={key} style={{ marginTop: '.5rem' }}>
                <input
                  name={code}
                  id={`${code}-${key}`}
                  type="checkbox"
                  value={key}
                  className="cb-f"
                  checked={values.indexOf(key) !== -1}
                  onChange={onChange}
                />
                <label className="cb-l" htmlFor={`${code}-${key}`}>
                  {name}
                </label>
              </div>
            )}
          </div>
        )}
      </div>

      {renderValidationMessages(question)}
    </div>
  );
}

function renderSingleValueListAttribute(question, onChange) {
  const { description, required, code, value, valueOptions } = question;

  const options = [...(valueOptions || [])];

  if (!required) {
    options.push({ key: '', name: 'No Answer' });
  }

  const lists = maybeSplitList(options, 5);
  const labelId = `${code}-label`;

  return (
    <div role="radiogroup" aria-labelledby={labelId}>
      <div className="m-v300">
        <label className="txt-l" id={labelId}>
          {description} {maybeRenderRequired(required)}
        </label>
      </div>
      <div className="g">
        {lists.map((list, i) =>
          <div
            className={lists.length === 1 ? 'g--12' : 'g--6'}
            key={`list-${i}`}
          >
            {list.map(({ key, name }) =>
              <div className="ra" key={key}>
                <input
                  name={code}
                  id={`${code}-${key}`}
                  type="radio"
                  value={key || '--no-answer-key--'}
                  className="ra-f"
                  checked={value === key || (key === '' && value === null)}
                  onChange={onChange}
                />
                <label className="ra-l" htmlFor={`${code}-${key}`}>
                  {name}
                </label>
              </div>
            )
            // blank span below to be the :last-child so that all the labels have a bottom margin
            }
            <span />
          </div>
        )}
      </div>

      {renderValidationMessages(question)}
    </div>
  );
}

@observer
export default class AttributeField extends React.Component<Props> {
  @action.bound
  onChange(ev: SyntheticInputEvent<>) {
    const { question } = this.props;
    const { value } = ev.target;

    if (question.type === 'NUMBER') {
      question.value = value.replace(/\D+/g, '');
    } else if (question.type === 'DATE') {
      if (value) {
        const datePieces = value.split('/');
        question.value = `${datePieces[2] || ''}-${datePieces[0] ||
          ''}-${datePieces[1] || ''}`;
      } else {
        question.value = '';
      }
    } else {
      question.value = value === '--no-answer-key--' ? '' : value;
    }
  }

  @action.bound
  onCheckbox(ev: SyntheticInputEvent<>) {
    const { question } = this.props;
    question.value = (ev.target.checked || false).toString();
  }

  @action.bound
  onMultivalueList(ev: SyntheticInputEvent<>) {
    const { question } = this.props;
    const values = currentValueAsArray(question.value);

    if (ev.target.checked) {
      question.value = [...values, ev.target.value];
    } else {
      question.value = values.filter(v => v !== ev.target.value);
    }
  }

  render() {
    const { question } = this.props;

    switch (question.type) {
      case 'BOOLEAN_CHECKBOX':
        return renderCheckbox(question, this.onCheckbox);
      case 'INFORMATIONAL':
        return renderInformationalAttribute(question);
      case 'DATETIME':
      case 'DATE':
        return <AttributeDateField question={question} />;
      case 'STRING':
        return renderStringAttribute(question, this.onChange);
      case 'NUMBER':
        return renderNumberAttribute(question, this.onChange);
      case 'TEXT':
        return renderTextAttribute(question, this.onChange);
      case 'SINGLEVALUELIST':
        return renderSingleValueListAttribute(question, this.onChange);
      case 'MULTIVALUELIST':
        return renderMultiValueListAttribute(question, this.onMultivalueList);
      default:
        return null;
    }
  }
}
