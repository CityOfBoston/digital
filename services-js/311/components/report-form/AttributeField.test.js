// @flow

import React from 'react';
import renderer from 'react-test-renderer';

import AttributeField from './AttributeField';

test('Text', () => {
  const component = renderer.create(
    <AttributeField
      attribute={{
        required: false,
        type: 'TEXT',
        code: 'ST-CMTS',
        description: 'Please provide any other relevant information:',
        values: null,
      }}
    />,
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('Informational', () => {
  const component = renderer.create(
    <AttributeField
      attribute={{
        required: false,
        type: 'INFORMATIONAL',
        code: 'INFO-NEDRMV1',
        description: '**All needle pickup cases should be followed up with a phone call to one of the below agencies.**',
        values: null,
      }}
    />,
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('Picklist', () => {
  const component = renderer.create(
    <AttributeField
      attribute={{
        required: true,
        type: 'PICKLIST',
        code: 'SR-NEDRMV1',
        description: 'How many needles are at the location?',
        values: [{ key: 'One', name: 'One' }, { key: 'Two', name: 'Two' }, { key: 'Three', name: 'Three' }, { key: 'More than Three', name: 'More than Three' }],
      }}
    />,
  );
  expect(component.toJSON()).toMatchSnapshot();
});
