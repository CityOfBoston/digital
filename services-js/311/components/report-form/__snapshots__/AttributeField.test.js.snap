exports[`test Informational 1`] = `
<p>
  **All needle pickup cases should be followed up with a phone call to one of the below agencies.**
</p>
`;

exports[`test Picklist 1`] = `
<label>
  <p>
    How many needles are at the location?
  </p>
  <select
    name="SR-NEDRMV1">
    <option
      value="One">
      One
    </option>
    <option
      value="Two">
      Two
    </option>
    <option
      value="Three">
      Three
    </option>
    <option
      value="More than Three">
      More than Three
    </option>
  </select>
</label>
`;

exports[`test Text 1`] = `
<label>
  <p>
    Please provide any other relevant information:
     
  </p>
  <textarea
    className={
      Object {
        "data-css-1bm19ha": "",
      }
    } />
</label>
`;
